query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/QueryBuilderDsl.java

```diff
@@ -0,0 +1,829 @@
+import com.datastax.oss.driver.internal.querybuilder.select.TupleSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.TypeHintSelector;
+import com.datastax.oss.driver.internal.querybuilder.term.BinaryArithmeticTerm;
+import com.datastax.oss.driver.internal.querybuilder.term.FunctionTerm;
+import com.datastax.oss.driver.internal.querybuilder.term.OppositeTerm;
+import com.datastax.oss.driver.internal.querybuilder.term.TupleTerm;
+import com.datastax.oss.driver.internal.querybuilder.term.TypeHintTerm;
+import com.google.common.collect.Iterables;
+import java.util.Arrays;
+import java.util.Map;
+
+/** A Domain-Specific Language to build CQL queries using Java code. */
+public class QueryBuilderDsl {
+
+  /** The identifier of the built-in {@code writetime} function. */
+  public static CqlIdentifier WRITETIME = CqlIdentifier.fromCql("writetime");
+
+  /** The identifier of the built-in {@code ttl} function. */
+  public static CqlIdentifier TTL = CqlIdentifier.fromCql("ttl");
+
+  /** Starts a SELECT query for a qualified table. */
+  public static SelectFrom selectFrom(CqlIdentifier keyspace, CqlIdentifier table) {
+    return new DefaultSelect(keyspace, table);
+  }
+
+  /**
+   * Shortcut for {@link #selectFrom(CqlIdentifier, CqlIdentifier)
+   * selectFrom(CqlIdentifier.fromCql(keyspace), CqlIdentifier.fromCql(table))}
+   */
+  public static SelectFrom selectFrom(String keyspace, String table) {
+    return selectFrom(CqlIdentifier.fromCql(keyspace), CqlIdentifier.fromCql(table));
+  }
+
+  /** Starts a SELECT query for an unqualified table. */
+  public static SelectFrom selectFrom(CqlIdentifier table) {
+    return selectFrom(null, table);
+  }
+
+  /** Shortcut for {@link #selectFrom(CqlIdentifier) selectFrom(CqlIdentifier.fromCql(table))} */
+  public static SelectFrom selectFrom(String table) {
+    return selectFrom(CqlIdentifier.fromCql(table));
+  }
+
+  /** Starts a DELETE query for a qualified table. */
+  public static DeleteSelection deleteFrom(CqlIdentifier keyspace, CqlIdentifier table) {
+    return new DefaultDelete(keyspace, table);
+  }
+
+  /**
+   * Shortcut for {@link #deleteFrom(CqlIdentifier, CqlIdentifier)
+   * deleteFrom(CqlIdentifier.fromCql(keyspace), CqlIdentifier.fromCql(table))}
+   */
+  public static DeleteSelection deleteFrom(String keyspace, String table) {
+    return deleteFrom(CqlIdentifier.fromCql(keyspace), CqlIdentifier.fromCql(table));
+  }
+
+  /** Starts a DELETE query for an unqualified table. */
+  public static DeleteSelection deleteFrom(CqlIdentifier table) {
+    return deleteFrom(null, table);
+  }
+
+  /** Shortcut for {@link #deleteFrom(CqlIdentifier) deleteFrom(CqlIdentifier.fromCql(table))} */
+  public static DeleteSelection deleteFrom(String table) {
+    return deleteFrom(CqlIdentifier.fromCql(table));
+  }
+
+  /** Selects all columns, as in {@code SELECT *}. */
+  public static Selector getAll() {
```

Comment:
I agree that the current names suck (I have a todo item to revisit them). Using different classes is not user-friendly for two reasons:

* you have to go look for the right class, instead of having everything at your disposal with a single static import.
* static imports still clash, i.e. you can only static import QueryBuilderDsl, you have to qualify the others.

I was thinking of `getColumn` => `selectColumn`. That's a bit repetitive when you do `selectFrom("foo").selector(selectColumn("k"))`,  but in real life you'll never do that, "static" code will use the shortcut `selectFrom("foo").column("k")`, `selector()` will only be used when you do stuff dynamically.

Reply:
All things considered, I'm going to do a full 180 on this. My idea of changing the prefix to `select` clashes with `selectFrom`, which is really bad.

The "discoverability" issue can be mitigated by moving to the most logical interface (`Selector`, `Relation`, etc.) and proper documentation.

If we drop the prefixes, we lose the ability to statically import everything, but those methods will be used mostly in dynamic code, where it's not as important to be concise, and anyway `Selector.column()` is not much longer than `selectColumn()`.